### DISCUSSION TITLE
How does Mineflayer controls the agent via simulating the actions of the keyboard and mouse?

### DISCUSSION THREAD

**Discussion by GoingMyWay** [PrismarineJS/mineflayer#3109] [Q&A]
Mineflayer has many high-level actions, such as `bot.goTo` and `bot.craft`. I am curious about how does Mineflayer simulates the keyboard and mouse controls in the code?

**Comment by PondWader** [comment]
Mineflayer's physics and movements are built on top of https://github.com/prismarinejs/prismarine-physics

For `bot.craft` it is just sending the packets to interact with the crafting menu. 

You can take a look at the code in Mineflayer in src/plugins for most parts
